OAuth support

ob-stripe opened the pull request:
r? @brandur-stripe 
cc @stripe/api-libraries 

Adds native support for OAuth in the Java library.

The implementation is mostly based on the existing ones (cf. https://github.com/stripe/stripe-ruby/pull/523). One notable difference is that I split the existing `_request` method in `LiveStripeResponseGetter` into two different `_request` (for normal API requests) and `_oauthRequest` (for OAuth requests) methods, with the common code factorized into a new `_rawRequest` method.

I did this because if I had simply updated the `_request` method to handle OAuth errors (like I did in the other languages implementations), then I'd have to update every request method in every model to account for the new possible OAuth exceptions (even though in practice, OAuth exceptions will never occur in normal API requests, but the compiler has no way of knowing this).

There might be a smarter / more elegant way of handling this. Open to all suggestions here!

ob-stripe commented:
Tests are failing because of an unrelated issue. They should pass once #399 is merged and this PR is rebased.

brandur-stripe commented on src/main/java/com/stripe/model/oauth/TokenResponse.java:
```diff
@@ -0,0 +1,98 @@
+
+	/**
+	 * @deprecated Use `stripe_user_id` with the `Stripe-Account` header (https://stripe.com/docs/connect/authentication#stripe-account-header)
+	 */
```
Is the idea to bring in deprecated fields for older API versions? I wonder if we should wait until someone complains ... :)

ob-stripe replied:
Yeah, I was being conservative here, but since this is a new feature as far as the Java library is concerned, we might as well not declare the deprecated fields at all. I'll update the PR.

brandur-stripe commented on src/main/java/com/stripe/net/OAuth.java:
```diff
@@ -0,0 +1,82 @@
+							+ "https://dashboard.stripe.com/account/applications/settings, "
+							+ "after registering your account as a platform. See "
+							+ "https://stripe.com/docs/connect/standard-accounts for details, "
+							+ "or email support@stripe.com if you have any questions.",
```
I think we'll have to change this error message right?

ob-stripe replied:
It's the same error message as in the other language implementations (e.g. in [stripe-ruby](https://github.com/stripe/stripe-ruby/blob/a645a78cd028554165e44760d531f946db0ea1cb/lib/stripe/oauth.rb#L18-L24)). What do you think needs to be changed?

brandur-stripe replied:
@ob-stripe Sorry, I should have been more specific. The config is called `Stripe.clientId` in Java (not `Stripe.client_id`).

ob-stripe replied:
🤦‍♂️  Of course, thanks!

brandur-stripe commented on src/main/java/com/stripe/net/LiveStripeResponseGetter.java:
```diff
@@ -605,6 +653,33 @@ private static void handleAPIError(String rBody, int rCode, String requestId)
+			case "unsupported_response_type":
+				throw new UnsupportedResponseTypeException(code, description, requestId, rCode, null);
+			default:
+				throw new APIException(code, requestId, rCode, null);
```
Great use of `switch`!

brandur-stripe commented on src/main/java/com/stripe/net/LiveStripeResponseGetter.java:
```diff
@@ -321,6 +340,12 @@ static String createQuery(Map<String, Object> params)
 		String charge;
 	}
 
+	private static class OAuthError {
```
I think a short comment on this one indicating that it's used to receive an error before converting it to a more suitable user-facing exception would be helpful.

ob-stripe replied:
:+1:

brandur-stripe commented:
Amazing work OB, thanks so much! I left a couple comments, but this looks great to me.

ob-stripe commented:
Alright, I pushed a couple of commits:

- I removed the deprecated fields from the token response, and updated the test to use a model file containing a full token response (to ensure that the deprecated fields are silently ignored by the library)

- I added some Javadoc to all the OAuth methods and exception classes, including `OAuthException`, and added a short comment above `OAuthError` in `LiveStripeResponseGetter` as you suggested

ptal @brandur-stripe

brandur-stripe commented:
> Alright, I pushed a couple of commits:

Nice work again!

LGTM pending minor change above.

ptal @ob-stripe

ob-stripe commented:
@brandur-stripe I fixed the error messages and squashed the commits. ptal

brandur-stripe commented:
Released as 5.8.0.
